// fluent-cypher/lib/props.js
exports.formatParamsMap = (paramsMap) => {
	let propsString = ''

	if(Object.keys(paramsMap).length){	
		propsString = ' {'
		let propsArray = []
		for(let propKey in paramsMap){
			propsArray.push(`${propKey}:{${paramsMap[propKey]}}`)
		}
		propsString += propsArray.join(', ')
		propsString += '}'
	}

	return propsString
}

exports.formatProps = (element, props, operator = "=") => {
	var a = [];
	for (var key in props) {
		var val = props[key];
		if(isString(val))
			a.push(`${element.alias}.${key} ${operator} "${val}"`);
		else if(isArray(val))
			a.push(`${element.alias}.${key} ${operator} [${val}]`);
		else
			a.push(`${element.alias}.${key} ${operator} ${val}`);
	};

	return a.join(', ');
}

exports.formatOrderBy = (props) => {
	var a = [];

	props.forEach((orderByProp) => {
		if(isString(orderByProp)){
			a.push(orderByProp);
		}else if(isObject(orderByProp)){
			orderByProp.dir = orderByProp.dir || 'ASC'
			a.push(`${orderByProp.alias}.${orderByProp.prop} ${orderByProp.dir}`);
		}
	});

	return a.join(', ');
}

exports.formatPropKeys = (element, propKeys) => {
	var a = [];
	propKeys.forEach((key)=>{
		a.push(`${element.alias}.${key}`);
	})

	return a.join(', ');
}

exports.formatCreatedAt = (obj) => `${obj.alias}.createdAt = timestamp()`

exports.formatUpdatedAt = (obj) => `${obj.alias}.updatedAt = timestamp()`

exports.formatMatchedAt = (obj) => `${obj.alias}.matchedAt = timestamp()`

exports.formatMatchCount = (obj) => `${obj.alias}.matchCount = coalesce(${obj.alias}.matchCount, 1) + 1`

exports.formatPropRegex = (element, prop) => {
	if(!element[prop])
		throw "The element to search does not contain " + prop

	const escapeStringRegexp = require('escape-string-regexp')
	return `${element.alias}.${prop} =~ '(?i).*${escapeStringRegexp(element[prop])}.*'`
}
